elastic/kibana
# [Fleet] Allow packages to specify index privileges

## Summary

Closes https://github.com/elastic/kibana/issues/109047

This PR will not be ready to merge until we have tested it end to end.

Allow packages to specify a restricted set of index privileges as part of their manifest.  There was previously some stub code which handled `permissions` however this was never fully implemented, @axw proposed moving these to `elasticsearch.privileges.indices`.

if a package has elasticsearch.privileges.indices in their data stream definition then:
- remove any invalid or forbidden privileges e.g `delete` is not allowed
- add these privileges to the package policy saved object
- read these fields from the package policy saved object when generating the agent policy

## Testing Steps

First I had to modify a package to specify index privileges in its manifest, for testing i used the custom logs package, these steps won't be necessary once https://github.com/elastic/apm-server/pull/6139 is merged.

- in the integrations repo edit packages/log/data_stream/log/manifest.yml to include the following at the top level:

```
elasticsearch:
  privileges:
    indices: [auto_configure, create_doc, maintenance, monitor, read]
```
- in the package directory then run `elastic-package build`
- run `elastic-package stack up -v -d --services package-registry` to start the package registry in docker locally
- in kibana.dev.yml set `xpack.fleet.registryUrl: http://localhost:8080`

We can now test if these privileges are added to the API key when I install the integration (on mac):

- add the custom logs package to a policy
- in dev tools run `GET /.fleet-agents/_search`
- find an agent that is part of your test policy and copy the `default_api_key`
- in a terminal run `echo -n "<your token>" | base64`
- copy that output into the following curl command, note you may need to change the index pattern depending on the test data stream you have used:
```
curl "http://localhost:9200/_security/user/_has_privileges" \
-H "Connection: keep-alive" \
-H "Content-Type: application/json" \
-H "Authorization: ApiKey <your base64 encoded token>" \
-d '{ "index" : [ { "names": [ "logs-generic-default" ] , "privileges": ["auto_configure", "create_doc", "maintenance", "monitor", "read"] } ] }'
```
- the response should contain `"has_all_requested":true`
### Checklist

Delete any items that are not applicable to this PR.
